[RPC][Better Engineering] Consolidated all rpcAgentRunning atomic booleans

osalpekar opened the pull request:
Summary: Closes: https://github.com/pytorch/pytorch/issues/32963

Test Plan: build bot

Differential Revision: D20074714

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -230,6 +230,11 @@ class TORCH_API RpcAgent {
+  // Atomic boolean indicating whether this agent is running. It is set in the
+  // child class start() and unset in child class shutdown() and join(). It
+  // controls whether several background threads should be running.
+  std::atomic<bool> rpcAgentRunning_;
```
should we explicitly give it a false initial value?

pritamdamania87 replied:
Can be done in the constructor initializer list.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -234,18 +234,19 @@ void ProcessGroupAgent::sync() {
   {
     std::lock_guard<std::mutex> futureLock{futureMutex_};
-    rpcRunning_.store(true);
+    rpcAgentRunning_.store(true);
```
hmm, should this be set in `RpcAgent::start()` and move `RpcAgent::start` before kicking off the `futureTimeoutThread_ `?

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -234,18 +234,19 @@ void ProcessGroupAgent::sync() {
 void ProcessGroupAgent::start() {
   {
     std::lock_guard<std::mutex> futureLock{futureMutex_};
```
Why do we need this lock here? Does the following comment also apply here? Are we worrying about two threads, one calling `start` and another calling `shutdown`? I am unclear of how this could go wrong. 

```c++
      // If during evaluating the predicate, `::shutdown()` is called, then
      // the predicate missed the notification before it started waiting
      // on the cond var.
```

osalpekar replied:
Made all the requested changes except this one - running some tests internally, and will update with this change shortly!

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -189,7 +189,7 @@ class TORCH_API RpcAgent {
 
   // start accepting requests
-  virtual void start() = 0;
+  virtual void start();
```
Let's clearly state what is the contact for derived classes. In the current implementation, it looks like there are at least two requirements:

1. derived class must call start from `RpcAgent`
2. derived class must set `rpcAgentRuninng_` before calling `RpcAgent::start()`

We might be able to remove the 2nd if we consolidate the `rpcAgentRunning_` into `RpcAgent::start()`

pritamdamania87 replied:
I think a non-virtual `start` and a pure virtual `startImpl` that subclasses need to override seems like a cleaner contract.

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -230,6 +230,11 @@ class TORCH_API RpcAgent {
   std::atomic<std::chrono::milliseconds> rpcTimeout_;
   std::atomic<bool> profilingEnabled_;
   std::shared_ptr<TypeResolver> typeResolver_;
+  // Boolean that indicates whether RpcAgent is running.
```
this line of comment is redundant?

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -230,6 +230,11 @@ class TORCH_API RpcAgent {
   std::shared_ptr<TypeResolver> typeResolver_;
+  // Boolean that indicates whether RpcAgent is running.
+  // Atomic boolean indicating whether this agent is running. It is set in the
+  // child class start() and unset in child class shutdown() and join(). It
```
Why do we require it is set in the child class?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -234,18 +234,19 @@ void ProcessGroupAgent::sync() {
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
+  RpcAgent::start();
```
Should the order here be reversed? RpcAgent::start() should execute first and do the common initialization(retryThread and rpcAgentRunning_ = true) and then should call start for the implementation? Can do this by having start() only in RpcAgent and make a `startImpl` pure virtual function in RpcAgent.

mrshenli reviewed (approved):
LGTM! Thanks!

jjlilley commented:
Please ensure that fb-side ThriftRpcAgentTest pass before landing (currently broken).
Also, it's not clear that start->startImpl rename is necessary here?

osalpekar committed:
[RPC][Better Engineering] Consolidated all rpcAgentRunning atomic booleans (#33915)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33915

Closes: https://github.com/pytorch/pytorch/issues/32963

Test Plan: build bot

Differential Revision: D20074714

fbshipit-source-id: 80ee6e3e57afe4e9778c2d69d64a7632901a91dc
